algebras/algebra/src/main/scala/endpoints4s/algebra/WebSockets.scala

```diff
@@ -0,0 +1,64 @@
+package endpoints4s.algebra
+
+import endpoints4s.Tupler
+
+trait WebSockets extends WebsocketClientStream with WebsocketServerStream with Urls {
+
+  type WebSocket[A, B]
+
+  def webSocket[URLC, CSC, In, Out](
+      url: Url[URLC],
+      clientStream: WebsocketClientStream[CSC] = emptyWebsocketClientStream,
+      serverStream: WebsocketServerStream[Out] = emptyWebsocketServerStream
+  )(implicit
+      tupler: Tupler.Aux[URLC, CSC, In]
+  ): WebSocket[In, Out]
+}
+
+trait WebSocketsWithUpgrade
+    extends WebSockets
+    with WebSocketUpgradeRequest
+    with WebSocketUpgradeResponse {
+
+  type WebSocketWithUpgrade[A, B]
+
+  def webSocketWithUpgrade[URLC, CSC, URHC, SSS, URHS, URLCAndCSCTupled, In, Out](
+      url: Url[URLC],
+      clientStream: WebsocketClientStream[CSC] = emptyWebsocketClientStream,
+      upgradeRequestHeaders: RequestHeaders[URHC] = emptyRequestHeaders,
+      serverStream: WebsocketServerStream[SSS] = emptyWebsocketServerStream,
+      upgradeResponseHeaders: RequestHeaders[URHS] = emptyRequestHeaders
+  )(
+      tuplerURLCCSC: Tupler.Aux[URLC, CSC, URLCAndCSCTupled],
+      tuplerURLCAndCSCTupledURHC: Tupler.Aux[URLCAndCSCTupled, URHC, In],
+      tuplerSSSURHS: Tupler.Aux[SSS, URHS, Out]
+  ): WebSocketWithUpgrade[In, Out]
+}
+
+trait WebsocketClientStream {
```

Comment:
I guess you mean _request_ stream?

Reply:
Terms such as request and response are not really used for anything but the upgrade in WebSocket docs that I have read, but there is a clear definition of client and server, so I thought I would use those terms.